File: 1870-minimum-speed-to-arrive-on-time/1870-minimum-speed-to-arrive-on-time.cpp
class Solution {
    bool ispossible(vector<int> &dist, long long mid, double hour)
    {
        double h=0;
        int n=dist.size();
        for(int i=0;i<n-1;i++)
        {
            h+=(dist[i]+mid-1)/mid;
        }
        h+=(double)dist[n-1]/mid;
        if(h<=hour)
        {
            return true;
        }
        return false;
    }

public:
    int minSpeedOnTime(vector<int>& dist, double hour) {
        long long ans=-1;
        long long s=1;
        long long e=1e7;
        while(s<=e)
        {
            long long mid=s+(e-s)/2;
            if(ispossible(dist,mid,hour))
            {
                ans=mid;
                e=mid-1;
            }
            else
            {
                s=mid+1;
            }
        }
        return ans;
    }
};

File: 1870-minimum-speed-to-arrive-on-time/1870-minimum-speed-to-arrive-on-time.py
class Solution:
    def ispossible(self, dist: List[int], mid: int, hour: float) -> int:
        h=0
        n=len(dist)
        for i in range(n-1):
            h+=(dist[i]+mid-1)//mid
        h+=dist[n-1]/mid
        if(h<=hour):
            return True
        return False

    def minSpeedOnTime(self, dist: List[int], hour: float) -> int:
        ans=-1
        s=1
        e=int(1e7)
        while(s<=e):
            mid=s+(e-s)//2
            if(self.ispossible(dist,mid,hour)):
                ans=mid
                e=mid-1
            else:
                s=mid+1
        return ans